### PR TITLE
buildsystem: move user scripts to contrib directory

### PR DESCRIPTION
**Pull request by feckert**
The script directory is a mess from my point of view. There are scripts that can be called by the build system and scripts that can be called by the user. In order to tidy things up a bit, all scripts that affect the user should be moved to the `bin` directory. I then also can extend the $PATH so that I can also use the scripts more easily without always specifying the whole openwrt path.
The `scripts` directory should then only contain scripts that are used by the buildsystem it self.

- For this to work, however, the output of the build must be moved to another directory with this [commit](https://github.com/openwrt/openwrt/commit/dd9a3eeba0766db0f995c06ee669e141faaf8ae6). In the `bin` folder the ipkg files of the whole build are now. They should be move to `artifacts` (The name could be changed).

- The second [commit](https://github.com/openwrt/openwrt/commit/4bf1de80a3e933ccf9d70a564575a64d4221f17c) moves some user scripts to the new `bin` directory. In my view, this is an iterative process. I have now started to move the first ones.

- With the third [commit](https://github.com/openwrt/openwrt/commit/4a3bc5aaaa7ede34e9d785757bbe0d86ddc1f990), I also want to show that scripts that are only needed for some targets during the build process should also go into the target directory. This has the advantage that when deleting a target that is no longer supported by OpenWrt, these scripts are also deleted and do not rot in the script folder.

### CODE REVIEW

**Comment by wulfy23** [comment]
agree in theory... 2c...

- not sure about touching /bin too ubiquitous... leave and use something else (same goes for artifacts would prefer alt name / location for that)
- third patch also raises the question/possibility re: target/image/xyz scripts and whether these be also relocated to same or nearby alt dir (i.e. bootscript.scr )

**Comment by adschm** [comment]
I also like the general idea, as well as the basic idea of commit 3.

However, I don't like the choice of the /bin directory for storing non-build scripts.
I've only started to think about this, though, so I cannot provide a better idea yet.

**Comment by nbd168** [comment]
I think that your choices of directory names would have been better if we had started with those. Having said that, I really don't like making such changes now. In my opinion, it's going to cause a lot of confusion, make a lot of scripts suddenly OpenWrt version dependent, and break muscle memory of many developers (including myself). And all of this just for the sake of making the directory layout slightly cleaner? I don't think it's worth the cost at all.

**Comment by nbd168** [comment]
I do agree with commit 3 though.

**Comment by adschm** [comment]
@nbd168 That's why I wonder whether we could just find a directory name for the non-build scripts that does not collide with anything. There should not be too many dependencies relying on these?

**Comment by aparcar** [comment]
How about `misc` or `contrib` for convenience scripts? Or move all build scripts to `./include/scripts`?

**Comment by feckert** [comment]
Thanks to all for the feedback
I link the suggestion from @aparcar, that we move all build scripts to the  include folder.
After all, they are called up the `*.mk` files in this directory.
Scripts that could be called by the user could stay in the scripts directory.

@nbd168
I can understand your argument. But I've been doing it for a long time now and I still have problems understanding everything in the buildsystem and if you have not been there from the beginning and know the history of each script.
From my point of view, it makes it easier to get started with the whole build system if there are rules.

- Scripts that are used during the build should go to includes/scripts that are helper scripts that are only needed by the buildsystem
- Scripts that the user can execute should stay in the scripts directory.
- Scripts that only affect certain targets should go to `target/linux/<target>/include/` or `target/linux/<target>/<subtarget>/include/ ` directory. Those scripts ara also only call by the buildsystem.
- Scripts that only used by the user and also only related to some targets should go to `target/linux/<target>/scripts/` or `target/linux/<target>/<subtarget>/scripts/ ` directory

**Comment by aparcar** [comment]
> Scripts that only used by the user and also only related to some targets should go to target/linux/<target>/scripts/ or target/linux/<target>/<subtarget>/scripts/ directory

Can you give an example here? I think hiding user scripts in some sub folder is a bad idea since nobody will ever find them

**Comment by feckert** [comment]
I have now searched where the [SCRIPT_DIR](https://github.com/openwrt/openwrt/search?q=SCRIPT_DIR) is used everywhere.

And I have also seen that some scripts also use scripts from the SCRIPT_DIR.
https://github.com/openwrt/openwrt/blob/master/scripts/patch-specs.sh#L73

As I see it now, I think I have to agree with @nbd168 that it will be difficult and painfull to move the build script to an other location.

I would still suggest that we move scripts that we don't need during the build processto a different directory.
I like @arpacar suggestion to move them to `contrib` or `misc`.
We should also move scripts that are only relevant for certain targets during the build the target/linux directory.

In summary, I would now update the pull request and move the scripts that a user can use to `contrib` or `misc` and move the target build scripts.

**Comment by aparcar** [comment]
Looks fine now, I'm a bit concerned that it may break the buildbots, could you please check that?

**Comment by aparcar** [comment]
The *find targets* step uses the `dump-target-info.pl` from the scripts folder, see https://git.openwrt.org/?p=buildbot.git;a=blob;f=phase1/master.cfg;h=be3a8941bb81d26f7a31a55c8835f64399cf35f2;hb=HEAD#l302

**Comment by adschm** [comment]
I wonder whether we should just introduce subfolders to scripts. This would remove the naming problem, and the change would be immediately obvious to anybody looking at the code even without having the history ...

**Comment by feckert** [comment]
@adschm That is also a possibility.
What do you suggest? We already have the directories `config` and `flash`, how about `contrib`, `misc` or `bin` for scripts that should be called by the user?

@aparcar Thanks, I hadn't thought of that. That the buildbot also uses scripts from this directory and not via a make target.
